plugins/s3_auth/aws_auth_v4.cc

```diff
@@ -0,0 +1,730 @@
+    bool hostHeader        = (0 == HOST.compare(lowName));
+    if (!xAmzHeader && !contentTypeHeader && !hostHeader) {
+      /* Skip internal headers (starting with '@'*/
+      if ('@' == name[0] /* exclude internal headers */) {
+        continue;
+      }
+
+      /* @todo do better here, since iterating over the headers in ATS is known to be less efficient,
+       * come up with a better way if include headers set is non-empty */
+      bool include =
+        (!includeHeaders.empty() && includeHeaders.end() != includeHeaders.find(lowName)); /* requested to be included */
+      bool exclude =
+        (!excludeHeaders.empty() && excludeHeaders.end() != excludeHeaders.find(lowName)); /* requested to be excluded */
+
+      if ((includeHeaders.empty() && exclude) || (!includeHeaders.empty() && (!include || exclude))) {
+#ifdef AWS_AUTH_V4_DETAILED_DEBUG_OUTPUT
+        std::cout << "ignore header: " << String(name, nameLen) << std::endl;
+#endif
+        continue;
+      }
+    }
+
+    String lowercaseName  = lowercase(name, nameLen);
+    size_t trimValueLen   = 0;
+    const char *trimValue = trimWhiteSpaces(value, valueLen, trimValueLen);
+
+    signedHeadersSet.insert(lowercaseName);
+    headersMap[lowercaseName] = String(trimValue, trimValueLen);
+  }
+
+  for (StringSet::iterator it = signedHeadersSet.begin(); it != signedHeadersSet.end(); it++) {
+    sha256Update(&canonicalRequestSha256Ctx, *it);
+    sha256Update(&canonicalRequestSha256Ctx, ":");
+    sha256Update(&canonicalRequestSha256Ctx, headersMap[*it]);
+    sha256Update(&canonicalRequestSha256Ctx, "\n");
+  }
+  sha256Update(&canonicalRequestSha256Ctx, "\n");
+
+  for (StringSet::iterator it = signedHeadersSet.begin(); it != signedHeadersSet.end(); ++it) {
+    if (!signedHeaders.empty()) {
+      signedHeaders.append(";");
+    }
+    signedHeaders.append(*it);
+  }
+
+  sha256Update(&canonicalRequestSha256Ctx, signedHeaders);
+  sha256Update(&canonicalRequestSha256Ctx, "\n");
+
+  /* Hex(SHA256Hash(<payload>) (no new-line char at end)
+   * @TODO support non-empty content, i.e. POST */
+  String payloadSha256Hash = getPayloadSha256(signPayload);
+  sha256Update(&canonicalRequestSha256Ctx, payloadSha256Hash);
+
+  /* Hex(SHA256Hash(<CanonicalRequest>)) */
+  sha256Final(canonicalRequestSha256Hash, &canonicalRequestSha256Ctx);
+#ifdef AWS_AUTH_V4_DETAILED_DEBUG_OUTPUT
+  std::cout << "</CanonicalRequest>" << std::endl;
+#endif
+  return base16Encode((char *)canonicalRequestSha256Hash, SHA256_DIGEST_LENGTH);
+}
+
+/**
+ * @brief Default AWS entry-point host name to region based on (S3):
+ *
+ * @see http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region
+ * it is used to get the region programmatically  w/o configuration
+ * parameters and can (meant to) be overwritten if necessary.
+ */
+const StringMap
+createDefaultRegionMap()
+{
+  StringMap m;
+  /* us-east-2 */
+  m["s3.us-east-2.amazonaws.com"]           = "us-east-2";
```

Comment:
Interesting. Does this imply that we might have to update this as AWS adds more availability zones for S3? I see that the user can provide his own map, so maybe that's good enough? Should we have such a map file as an "example", which includes all of these mappings that are hardcoded here ?